src/async/GenericMuxedInterruptDistributor.cc

```diff
@@ -3,27 +3,718 @@
+
+  GenericMuxedInterruptDistributor::GenericMuxedInterruptDistributor(
+      boost::shared_ptr<SubDomain<std::nullptr_t>> parent, std::string registerPath,
+      std::bitset<GmidOptionCode::OPTION_CODE_COUNT> optionRegisterSettings)
+  : MuxedInterruptDistributor(parent), _path(registerPath.c_str()) {
+    // Set required registers
+    optionRegisterSettings.set(ISR);              // Ensure that the required option ISR is always set.
+    if(not optionRegisterSettings.test(IMaskR)) { // Ensure IMaskR or IER is on
+      optionRegisterSettings.set(IER);
+    }
+    // Note that we currently don't allowing IMaskR, but use of it gets caught later.
+
+    // Here we could explicitly note that we're ignoring IPR with optionRegisterSettings.reset(IPR)
+
+    /*----------------------------------------------------------------------------------------------------------------*/
+    steriliseOptionRegisterSettings(optionRegisterSettings, parent->getId());
+
+    /*----------------------------------------------------------------------------------------------------------------*/
+    _isr = _backend->getRegisterAccessor<uint32_t>(_path / getOptionRegisterStr(ISR), 1, 0, {});
+
+    _ierIsReallyImaskr = optionRegisterSettings.test(IMaskR);
+    _ier = _backend->getRegisterAccessor<uint32_t>(
+        _path / getOptionRegisterStr(_ierIsReallyImaskr ? IMaskR : IER), 1, 0, {});
+
+    // Set the clear/acknowledge register {ICR, IAR, ISR}
+    if(optionRegisterSettings.test(ICR)) { // Use ICR as the clear register
+      _icr = _backend->getRegisterAccessor<uint32_t>(_path / getOptionRegisterStr(ICR), 1, 0, {});
+    }
+    else if(optionRegisterSettings.test(IAR)) { // Use IAR as the clear register
+      _icr = _backend->getRegisterAccessor<uint32_t>(_path / getOptionRegisterStr(IAR), 1, 0, {});
+    }
+    else { // Use ISR to clear interrupts using its write 1 to clear feature
+      _icr = _backend->getRegisterAccessor<uint32_t>(_path / getOptionRegisterStr(ISR), 1, 0, {});
+    }
+
+    _hasMer = optionRegisterSettings.test(MER) or optionRegisterSettings.test(MIE) or optionRegisterSettings.test(GIE);
+    if(_hasMer) {
+      GmidOptionCode _optionMerMieGie =
+          optionRegisterSettings.test(MIE) ? MIE : (optionRegisterSettings.test(GIE) ? GIE : MER);
+      _mer = _backend->getRegisterAccessor<uint32_t>(_path / getOptionRegisterStr(_optionMerMieGie), 1, 0, {});
+    }
+
+    // steriliseOptionRegisterSettings ensures that either SIE and CIE are both enabled or neither are
+    _haveSieAndCie = optionRegisterSettings.test(SIE);
+    if(_haveSieAndCie) {
+      _sie = _backend->getRegisterAccessor<uint32_t>(_path / getOptionRegisterStr(SIE), 1, 0, {});
+      _cie = _backend->getRegisterAccessor<uint32_t>(_path / getOptionRegisterStr(CIE), 1, 0, {});
+    }
+
+    /*----------------------------------------------------------------------------------------------------------------*/
+    // Check register readability
```

Comment:
why are we only checking for readability, but not the write-to registers for writeability?

Reply:
You are right, this is incomplete. If we accidentally pick the wrong register, there will be a logic_error where it should not be. I am still uncertain whether to throw a logic error or a runtime error here, because we are checking the accessor, not the catalogue.

Reply:
Solved. Actually we only need readability for ISR, and writeability  for all others.
Checking the catalogue now, and throwing logic errors.